src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt

```diff
@@ -2231,107 +2240,111 @@ abstract class ManagedStrategy(
         return threadId in (0 ..< nThreads)
     }
 
-    /**
-     * Logs thread events such as thread switches and passed code locations.
-     */
-    private inner class TraceCollector {
-        private val _trace = mutableListOf<TracePoint>()
-        val trace: List<TracePoint> = _trace
-        private var spinCycleStartAdded = false
-
-        private val spinCycleMethodCallsStackTraces: MutableList<List<CallStackTraceElement>> = mutableListOf()
-
-        fun newSwitch(iThread: Int, reason: SwitchReason, beforeMethodCallSwitch: Boolean = false) {
-            if (reason == SwitchReason.ActiveLock) {
-                afterSpinCycleTraceCollected(
-                    trace = _trace,
-                    callStackTrace = callStackTrace[iThread]!!,
-                    spinCycleMethodCallsStackTraces = spinCycleMethodCallsStackTraces,
-                    iThread = iThread,
-                    currentActorId = currentActorId[iThread]!!,
-                    beforeMethodCallSwitch = beforeMethodCallSwitch
-                )
-            }
-            val callStackTrace = when (reason) {
-                SwitchReason.Suspended -> suspendedFunctionsStack[iThread]!!.reversed()
-                else -> callStackTrace[iThread]!!
+    // == TRACE COLLECTOR EXTENSION METHODS ==
+    private fun TraceCollector.passCodeLocation(tracePoint: TracePoint?) {
+        if (tracePoint !is SectionDelimiterTracePoint && tracePoint?.isActorMethodCallTracePoint() == false) {
+            checkActiveLockDetected()
+        }
+
+        if (tracePoint != null) {
+            // tracePoint can be null here if trace is not available, e.g. in case of suspension
+            passCodeLocationInternal(tracePoint)
+
+            if (!tracePoint.isActorMethodCallTracePoint()) {
+                setBeforeEventId(tracePoint)
             }
-            _trace += SwitchEventTracePoint(
-                iThread = iThread,
-                actorId = currentActorId[iThread]!!,
-                reason = reason,
-                callStackTrace = callStackTrace,
-            )
-            spinCycleStartAdded = false
         }
+    }
+
+    private fun TraceCollector?.newSwitch(reason: SwitchReason, beforeMethodCallSwitch: Boolean) {
```

Comment:
Let's move `?` to the call-site.

```suggestion
    private fun TraceCollector.newSwitch(reason: SwitchReason, beforeMethodCallSwitch: Boolean) {
```